File: 32--Save-the-Clock-Tower/HighVoltage/High Voltage/VoltageCalculatorTableViewController.m
//
//  VoltageCalculatorTableViewController.m
//  High Voltage
//
//  Created by Ben Gohlke on 3/10/15.
//  Copyright (c) 2015 The Iron Yard. All rights reserved.
//

#import "VoltageCalculatorTableViewController.h"
#import "ValueTypesTableViewController.h"

#import "ElectricityConversions.h"

@interface VoltageCalculatorTableViewController () <UITextFieldDelegate, UIPopoverPresentationControllerDelegate>
{
    UITextField *currentTextField;
    UITextField *resistanceTextField;
    UITextField *voltsTextField;
    UITextField *powerTextField;
    
    ElectricityConversions *converter;
    
    NSMutableArray *tableData;
    NSMutableDictionary *valueTypes;
}

@property (weak, nonatomic) IBOutlet UIBarButtonItem *clearBarButtonItem;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *addValueTypeBarButtonItem;

- (IBAction)clearTransaction:(UIBarButtonItem *)sender;

@end

@implementation VoltageCalculatorTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    tableData = [[NSMutableArray alloc] init];
    valueTypes = [[NSMutableDictionary alloc] initWithObjects:@[@"CurrentCell", @"ResistanceCell", @"VoltsCell", @"PowerCell"] forKeys:@[@"Amps", @"Ohms", @"Volts", @"Watts"]];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [tableData count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    NSString *title;
    
    if ([tableData count] > 0)
    {
        title = @"Enter two values";
    }
    else
    {
        title = @"";
    }
    
    return title;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *identifier = [tableData objectAtIndex:indexPath.row];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier forIndexPath:indexPath];
    
    UITextField *textField = (UITextField *)[cell viewWithTag:1];
    textField.text = @"";
    [textField setUserInteractionEnabled:!converter.allValuesFound];
    
    if ([identifier isEqualToString:@"CurrentCell"])
    {
        currentTextField = textField;
        if (converter.ampsString)
        {
            [textField setText:converter.ampsString];
        }
    }
    else if ([identifier isEqualToString:@"ResistanceCell"])
    {
        resistanceTextField = textField;
        if (converter.ohmsString)
        {
            [textField setText:converter.ohmsString];
        }
    }
    else if ([identifier isEqualToString:@"VoltsCell"])
    {
        voltsTextField = textField;
        if (converter.voltsString)
        {
            [textField setText:converter.voltsString];
        }
    }
    else if ([identifier isEqualToString:@"PowerCell"])
    {
        powerTextField = textField;
        if (converter.wattsString)
        {
            [textField setText:converter.wattsString];
        }
    }
    
    NSString *key = [valueTypes allKeysForObject:identifier][0];
    [valueTypes removeObjectForKey:key];
    
    [textField becomeFirstResponder];
    
    return cell;
}

#pragma mark - TextField delegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    BOOL rc = NO;
    
    if (![textField.text isEqualToString:@""])
    {
        rc = YES;
        if (textField == currentTextField)
        {
            converter.ampsString = textField.text;
        }
        else if (textField == resistanceTextField)
        {
            converter.ohmsString = textField.text;
        }
        else if (textField == voltsTextField)
        {
            converter.voltsString = textField.text;
        }
        else if (textField == powerTextField)
        {
            converter.wattsString = textField.text;
        }
    }
    
    if (rc)
    {
        [textField resignFirstResponder];
    }
    
//    [converter findOtherValuesIfPossible];
    
    return rc;
}

#pragma mark - ValuesTypeViewController delegate

- (void)valueTypeWasChosen:(NSString *)chosenValueType
{
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
    
    if (!converter)
    {
        converter = [[ElectricityConversions alloc] init];
        converter.delegate = self;
    }
    
    NSString *cellIdentifier = valueTypes[chosenValueType];
    [tableData addObject:cellIdentifier];
    if ([tableData count] == 2)
    {
        [self.addValueTypeBarButtonItem setEnabled:NO];
    }
    
    NSUInteger row = [tableData indexOfObject:cellIdentifier];
    [self.tableView insertRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:row inSection:0]] withRowAnimation:UITableViewRowAnimationAutomatic];
}

#pragma mark - ElectricityConversions delegate

- (void)valuesWereCalculated
{
    [self resetValueTypesDictionary];
    
    if (!voltsTextField)
    {
        NSString *cellIdentifier = valueTypes[@"Volts"];
        [tableData addObject:cellIdentifier];
    }
    if (!currentTextField)
    {
        NSString *cellIdentifier = valueTypes[@"Amps"];
        [tableData addObject:cellIdentifier];
    }
    if (!resistanceTextField)
    {
        NSString *cellIdentifier = valueTypes[@"Ohms"];
        [tableData addObject:cellIdentifier];
    }
    if (!powerTextField)
    {
        NSString *cellIdentifier = valueTypes[@"Watts"];
        [tableData addObject:cellIdentifier];
    }
    
    [self.tableView reloadData];
}

#pragma mark - Action Handlers

- (IBAction)clearTransaction:(UIBarButtonItem *)sender
{
    [tableData removeAllObjects];
    [self.addValueTypeBarButtonItem setEnabled:YES];
    converter = nil;
    voltsTextField = nil;
    currentTextField = nil;
    resistanceTextField = nil;
    powerTextField = nil;
    [self resetValueTypesDictionary];
    [self.tableView reloadData];
}

#pragma mark - Navigation
 
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ShowValueTypePopoverSegue"])
    {
        ValueTypesTableViewController *valueTypesVC = (ValueTypesTableViewController *)[segue destinationViewController];
        valueTypesVC.delegate = self;
        valueTypesVC.popoverPresentationController.delegate = self;
        float contentHeight = 44.0f * [valueTypes count];
        valueTypesVC.preferredContentSize = CGSizeMake(100.0f, contentHeight);
        valueTypesVC.valueTypes = [valueTypes allKeys];
    }
}

#pragma mark - Private methods

- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller
{
    return UIModalPresentationNone;
}

- (void)resetValueTypesDictionary
{
    [valueTypes removeAllObjects];
    [valueTypes addEntriesFromDictionary:@{@"Amps": @"CurrentCell", @"Ohms": @"ResistanceCell", @"Volts": @"VoltsCell", @"Watts": @"PowerCell"}];
}
 

@end

File: 16--Come-About/SwiftVoltage/High Voltage/VoltageCalculatorTableViewController.swift
//
//  VoltageCalculatorTableViewController.swift
//  High Voltage
//
//  Created by Ben Gohlke on 7/26/15.
//  Copyright (c) 2015 The Iron Yard. All rights reserved.
//

import UIKit

protocol ValueTypesViewControllerDelegate
{
    func valueTypeWasChosen(chosenValueType: String)
}

protocol ElectricityConversionsDelegate
{
    func valuesWereCalculated()
}

class VoltageCalculatorTableViewController: UITableViewController, ValueTypesViewControllerDelegate, ElectricityConversionsDelegate, UITextFieldDelegate, UIPopoverPresentationControllerDelegate
{
    
    @IBOutlet weak var clearBarButtonItem: UIBarButtonItem!
    @IBOutlet weak var addValueTypeBarButtonItem: UIBarButtonItem!
    
    var currentTextField: UITextField?
    var resistanceTextField: UITextField?
    var voltageTextField: UITextField?
    var powerTextField: UITextField?
    
    var converter: ElectricityConversions?
    
    var tableData = [String]()
    var valueTypes: Dictionary<String, String> = ["Amps": "CurrentCell", "Ohms": "ResistanceCell", "Volts": "VoltageCell", "Watts": "PowerCell"]
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        self.title = "High Voltage"
    }

    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int
    {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return tableData.count
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String?
    {
        if tableData.count > 0
        {
            return "Enter two values"
        }
        else
        {
            return ""
        }
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        let identifier = tableData[indexPath.row]
        
        let cell = tableView.dequeueReusableCellWithIdentifier(identifier, forIndexPath: indexPath) 

        let textField = cell.viewWithTag(1) as! UITextField
        textField.text = ""
        textField.delegate = self
        textField.userInteractionEnabled = !converter!.allValuesFound
        
        switch identifier
        {
        case "CurrentCell":
            currentTextField = textField
            if converter?.ampsString != ""
            {
                textField.text = converter?.ampsString
            }
            
        case "ResistanceCell":
            resistanceTextField = textField
            if converter?.ohmsString != ""
            {
                textField.text = converter?.ohmsString
            }
            
        case "PowerCell":
            powerTextField = textField
            if converter?.wattsString != ""
            {
                textField.text = converter?.wattsString
            }
            
        case "VoltageCell":
            voltageTextField = textField
            if converter?.voltsString != ""
            {
                textField.text = converter?.voltsString
            }
            
        default:
            print("")
        }

        let keys = (valueTypes as NSDictionary).allKeysForObject(identifier) as! [String]
        let keyToRemove = keys[0]
        valueTypes.removeValueForKey(keyToRemove)

        textField.becomeFirstResponder()

        return cell
    }
    
    // MARK: - UITextField delegate
    
    func textFieldShouldReturn(textField: UITextField) -> Bool
    {
        var rc = false
        if textField.text != ""
        {
            rc = true
            if textField == currentTextField
            {
                converter?.ampsString = textField.text!
            }
            if textField == resistanceTextField
            {
                converter?.ohmsString = textField.text!
            }
            if textField == voltageTextField
            {
                converter?.voltsString = textField.text!
            }
            if textField == powerTextField
            {
                converter?.wattsString = textField.text!
            }
        }
        
        if rc
        {
            textField.resignFirstResponder()
        }
        
        converter?.findOtherValuesIfPossible()
        
        return rc
    }
    
    // MARK: - ValueTypesViewController delegate
    
    func valueTypeWasChosen(chosenValueType: String)
    {
        navigationController?.dismissViewControllerAnimated(true, completion: nil)
        if converter == nil
        {
            converter = ElectricityConversions()
            converter?.delegate = self
        }
        
        let cellIdentifier = valueTypes[chosenValueType]
        tableData.append(cellIdentifier!)
        if tableData.count == 2
        {
            addValueTypeBarButtonItem.enabled = false
        }
        
        let row = (tableData as NSArray).indexOfObject(cellIdentifier!)
        tableView.insertRowsAtIndexPaths([NSIndexPath(forRow: row, inSection: 0)], withRowAnimation: .Automatic)
    }
    
    // MARK: - ElectricityConversions delegate
    
    func valuesWereCalculated()
    {
        resetValueTypesDictionary()
        
        if voltageTextField == nil
        {
            let cellIdentifier = valueTypes["Volts"]
            tableData.append(cellIdentifier!)
        }
        if currentTextField == nil
        {
            let cellIdentifier = valueTypes["Amps"]
            tableData.append(cellIdentifier!)
        }
        if resistanceTextField == nil
        {
            let cellIdentifier = valueTypes["Ohms"]
            tableData.append(cellIdentifier!)
        }
        if powerTextField == nil
        {
            let cellIdentifier = valueTypes["Watts"]
            tableData.append(cellIdentifier!)
        }
        
        tableView.reloadData()
    }
    
    // MARK: - Action Handlers
    
    @IBAction func clearTransaction(sender: UIBarButtonItem)
    {
        tableData.removeAll(keepCapacity: true)
        addValueTypeBarButtonItem.enabled = true
        converter = nil
        voltageTextField = nil
        currentTextField = nil
        resistanceTextField = nil
        powerTextField = nil
        resetValueTypesDictionary()
        tableView.reloadData()
    }

    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?)
    {
        if segue.identifier == "ShowValueTypePopoverSegue"
        {
            let valueTypesVC = segue.destinationViewController as! ValueTypesTableViewController
            valueTypesVC.delegate = self
            valueTypesVC.popoverPresentationController?.delegate = self
            let contentHeight = 44.0 * CGFloat(valueTypes.count)
            valueTypesVC.preferredContentSize = CGSizeMake(CGFloat(100.0), contentHeight)
            valueTypesVC.valueTypes = Array(valueTypes.keys)
        }
    }
    
    // MARK: - UIPopoverPresentation delegate
    
    func adaptivePresentationStyleForPresentationController(controller: UIPresentationController) -> UIModalPresentationStyle
    {
        return .None
    }
    
    // MARK: - Private methods
    
    func resetValueTypesDictionary()
    {
        valueTypes.removeAll(keepCapacity: true)
        valueTypes = ["Amps": "CurrentCell", "Ohms": "ResistanceCell", "Volts": "VoltageCell", "Watts": "PowerCell"]
    }
}
